File: renderers/web_core/src/v1_0/catalog/catalog-definition.ts
/**
 * Copyright 2026 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * @fileoverview Zod schemas and types for A2UI v1.0 Catalog Definitions.
 */

import {z} from 'zod';
import {isValidUax31Identifier} from '../validating/uax31.js';

/**
 * Component definition schema including composition rules (allowedParents, allowedChildren).
 */
export const ComponentDefinitionSchema = z.object({
  allowedParents: z.array(z.string()).optional(),
  allowedChildren: z.array(z.string()).optional(),
  metadata: z.record(z.unknown()).optional(),
});

export type ComponentDefinition = z.infer<typeof ComponentDefinitionSchema>;

/**
 * Function definition schema including returnType, callableFrom, and activation requirements.
 */
export const FunctionDefinitionSchema = z.object({
  returnType: z.enum(['string', 'number', 'boolean', 'array', 'object', 'any', 'void']),
  callableFrom: z
    .enum(['rendererOnly', 'agentOnly', 'rendererOrAgent'])
    .optional()
    .default('rendererOnly'),
  requiresUserActivation: z.boolean().optional().default(false),
  description: z.string().optional(),
  properties: z.record(z.unknown()).optional(),
});

export type FunctionDefinition = z.infer<typeof FunctionDefinitionSchema>;

/**
 * Catalog definition schema for A2UI v1.0.
 * Map-based functions and components where keys must satisfy UAX #31.
 */
export const CatalogDefinitionSchema = z.object({
  protocolVersion: z.string().optional().default('1.0'),
  catalogId: z.string(),
  title: z.string().optional(),
  description: z.string().optional(),
  instructions: z.string().optional(),
  components: z
    .record(
      z
        .string()
        .refine(isValidUax31Identifier, {
          message: 'Component name must be a valid UAX #31 identifier',
        })
        .refine(name => name !== 'Surface', {
          message: 'Component name cannot be Surface',
        }),
      ComponentDefinitionSchema,
    )
    .optional(),
  functions: z
    .record(
      z.string().refine(isValidUax31Identifier, {
        message: 'Function name must be a valid UAX #31 identifier',
      }),
      FunctionDefinitionSchema,
    )
    .optional(),
});

export type CatalogDefinition = Omit<z.infer<typeof CatalogDefinitionSchema>, 'protocolVersion'> & {
  protocolVersion?: string;
};

File: agent_sdks/python/a2ui_core/src/a2ui/core/catalog/catalog_definition.py
"""Pydantic models and schemas for A2UI v1.0 Catalog Definitions."""

from typing import Dict, List, Literal, Optional, Any
from pydantic import BaseModel, ConfigDict, Field, field_validator
from a2ui.core.validating.uax31 import is_valid_uax31_identifier


class ComponentDefinition(BaseModel):
    """Component definition model including composition constraints."""

    model_config = ConfigDict(extra="forbid")

    allowed_parents: Optional[List[str]] = Field(
        default=None,
        alias="allowedParents",
        description="List of allowed parent component types.",
    )
    allowed_children: Optional[List[str]] = Field(
        default=None,
        alias="allowedChildren",
        description="List of allowed child component types.",
    )
    metadata: Optional[Dict[str, Any]] = Field(
        default=None,
        description="Optional static metadata.",
    )


class FunctionDefinition(BaseModel):
    """Function definition model including returnType and callableFrom."""

    model_config = ConfigDict(extra="forbid")

    return_type: Literal[
        "string", "number", "boolean", "array", "object", "any", "void"
    ] = Field(
        alias="returnType",
        description="The type of value this function returns.",
    )
    callable_from: Literal["rendererOnly", "agentOnly", "rendererOrAgent"] = Field(
        default="rendererOnly",
        alias="callableFrom",
        description="Specifies where this function can be invoked from.",
    )
    requires_user_activation: bool = Field(
        default=False,
        alias="requiresUserActivation",
        description="Whether user activation context is required.",
    )
    description: Optional[str] = None
    properties: Optional[Dict[str, Any]] = None


class CatalogDefinition(BaseModel):
    """Catalog definition for A2UI v1.0."""

    model_config = ConfigDict(extra="forbid")

    protocol_version: Optional[str] = Field(
        default="1.0",
        alias="protocolVersion",
        description="A2UI specification version of this catalog.",
    )
    catalog_id: str = Field(
        alias="catalogId",
        description="Unique identifier for this catalog.",
    )
    title: Optional[str] = None
    description: Optional[str] = None
    instructions: Optional[str] = None
    components: Optional[Dict[str, ComponentDefinition]] = None
    functions: Optional[Dict[str, FunctionDefinition]] = None

    @field_validator("components", mode="after")
    @classmethod
    def validate_component_names(
        cls, v: Optional[Dict[str, ComponentDefinition]]
    ) -> Optional[Dict[str, ComponentDefinition]]:
        """Ensures all component names are valid UAX #31 identifiers and not 'Surface'."""
        if v is not None:
            for name in v.keys():
                if name == "Surface":
                    raise ValueError("Component name cannot be 'Surface'")
                if not is_valid_uax31_identifier(name):
                    raise ValueError(
                        f"Component name '{name}' must be a valid UAX #31 identifier"
                    )
        return v

    @field_validator("functions", mode="after")
    @classmethod
    def validate_function_names(
        cls, v: Optional[Dict[str, FunctionDefinition]]
    ) -> Optional[Dict[str, FunctionDefinition]]:
        """Ensures all function names are valid UAX #31 identifiers."""
        if v is not None:
            for name in v.keys():
                if not is_valid_uax31_identifier(name):
                    raise ValueError(
                        f"Function name '{name}' must be a valid UAX #31 identifier"
                    )
        return v
